core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,407 @@
+            return false;
+        }
+        wroteAny = true;
+        queue.add(o);
+        tryNotifyEvents();
+        return true;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenConsumed() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        close0(SUCCESSFUL_CLOSE);
+    }
+
+    @Override
+    public void close(Throwable cause) {
+        requireNonNull(cause, "cause");
+        close0(new CloseEvent(cause));
+    }
+
+    private void close0(CloseEvent closeEvent) {
+        assert eventLoop.inEventLoop();
+
+        if (state != State.OPEN) {
+            return;
+        }
+        state = State.CLOSED;
+
+        queue.add(closeEvent);
+        tryNotifyEvents();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
+    }
+
+    @Override
+    public long demand() {
+        return demand;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+        if (eventLoop.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            eventLoop.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<?> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        assert eventLoop.inEventLoop();
+        if (subscription != null) {
+            notifySubscribeAndAbort(NoopSubscription.get(), subscriber,
+                                    abortedOrLate(subscription.subscriber()), executor);
+            return;
+        }
+        subscription = new SubscriptionImpl(this, subscriber, executor);
+        notifySubscribe0(subscription, subscriber, executor);
+        tryNotifyEvents();
+    }
+
+    @Override
+    public void abort() {
+        if (eventLoop.inEventLoop()) {
+            abort0(ABORTED_CLOSE);
+        } else {
+            eventLoop.execute(() -> abort0(ABORTED_CLOSE));
+        }
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (eventLoop.inEventLoop()) {
+            abort0(new CloseEvent(cause));
+        } else {
+            eventLoop.execute(() -> abort0(new CloseEvent(cause)));
+        }
+    }
+
+    private void abort0(CloseEvent closeEvent) {
+        assert eventLoop.inEventLoop();
+        if (subscription == null) {
+            subscription = new SubscriptionImpl(this, AbortingSubscriber.get(closeEvent.cause()),
+                                                ImmediateEventExecutor.INSTANCE);
```

Comment:
How about implementing `Subscription` directly by `EventLoopStreamMessage` to avoid allocating an additional object?